File: src/data_structure/linked_list/print_in_reverse.java
/*
  Insert Node at the end of a linked list 
  head pointer input could be NULL as well for empty list
  Node is defined as 
  class Node {
     int data;
     Node next;
  }
*/
    // This is a "method-only" submission. 
    // You only need to complete this method. 

void ReversePrint(Node head) {
    if (head == null) {
        return;
    }
    if (head.next == null) {
        System.out.println(head.data);
        return;
    }
    Stack<Node> queue = new Stack<Node>();
    while (head != null) {
        queue.push(head);
        head = head.next;
    }
    while (!queue.isEmpty()) {
        System.out.println(queue.pop().data);
    }
}

File: src/data_structure/linked_list/print_in_reverse.cpp
/*
  Print elements of a linked list in reverse order as standard output
  head pointer could be NULL as well for empty list
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/
void ReversePrint(Node *head)
{
    if (head == NULL)
    {
        return;
    }
    if (head->next == NULL) {
        cout << head->data << endl;
        return;
    }
    Node *p, *q, *r;
    p = head;
    q = head->next;
    p->next = NULL;

    do
    {
        r = q->next;
        q->next = p;
        p = q;
        q = r;
    } while (r != NULL);

    // at this moment, p is header.
    while (p != NULL) {
        cout << p->data << endl;
        p = p->next;
    }
}

